eng/pipelines/optprof.yml

```
@@ -132,14 +140,43 @@ stages:
       displayName: 'Set RunSettingsURI variable'
       name: SetRunSettingsURI
     - download: ComponentBuildUnderTest
-      artifact: MicroBuildOutputs
-      patterns: '**\BootstrapperInfo.json'
-      displayName: Download Bootstrapper Information
+      artifact: VS15
+      patterns: "VS15/Microsoft.*"
+      displayName: 'Download VS15 files from official build'
+    - task: MicroBuildSigningPlugin@1
+      inputs:
+        signType: "real"
+        esrpSigning: "true"
+      displayName: "Install Signing Plugin"
+    - task: MicroBuildSwixPlugin@4
+      displayName: "Install Swix Plugin"
+    - task: MicroBuildBuildVSBootstrapper@2
+      displayName: 'Build a Visual Studio bootstrapper'
+      inputs:
+        channelName: "$(VsTargetChannel)"
+        vsMajorVersion: "$(VsTargetMajorVersion)"
+        manifests: '$(Pipeline.Workspace)\ComponentBuildUnderTest\VS15\Microsoft.VisualStudio.NuGet.Core.vsman'
+        outputFolder: '$(Pipeline.Workspace)\ComponentBuildUnderTest\VS15'
+    - task: PublishPipelineArtifact@1
+      displayName: 'Publish BootstrapperInfo.json as a pipeline artifact'
+      inputs:
+        targetPath: $(Build.StagingDirectory)\MicroBuild\Output
+        artifactName: 'MicroBuildOutputs'
+        artifactType: 'pipeline'
     - task: PowerShell@2
       displayName: Set 'VisualStudio.InstallationUnderTest.BootstrapperURL'
       inputs:
         filePath: $(DartLab.Path)\Scripts\VisualStudio\Bootstrapper\Get-BootstrapperURL.ps1
-        arguments: -BootstrapperInfoJsonURI '$(Pipeline.Workspace)\ComponentBuildUnderTest\MicroBuildOutputs\BootstrapperInfo.json' -VSBranch '$(VSBranch)' -OutVariableName 'VisualStudio.InstallationUnderTest.BootstrapperURL'
+        arguments: -BootstrapperInfoJsonURI '$(Build.StagingDirectory)\MicroBuild\Output\BootstrapperInfo.json' -VSBranch '$(VSBranch)' -OutVariableName 'VisualStudio.InstallationUnderTest.BootstrapperURL'
+    - task: artifactDropTask@0
+      displayName: "Upload VSTS Drop"
+      inputs:
+        dropServiceURI: 'https://devdiv.artifacts.visualstudio.com'
+        buildNumber: 'Products/$(System.TeamProject)/$(Build.Repository.Name)/$(Build.SourceBranchName)/$(Build.BuildNumber)'
```

This might clash with an official build. Well, if the build numbers used the same format, which I don't think is, so perhaps it's unintentially unique.  Fortunately, drops are write-once, so if this pipeline tried to overwrite an official build drop, this pipeline would fail.

```suggestion
        buildNumber: 'OptProf/$(System.TeamProject)/$(Build.Repository.Name)/$(Build.SourceBranchName)/$(Build.BuildNumber)'
```

However, I'm not sure what other consequences there are. Like, how do we tell the bootstrapper task that we want to use a different prefix?

---

`buildNumber` will not clash with the official build because the last parameter `$(Build.BuildNumber)` is different between official build and optprof pipeline. As you know for official build, we generate build number from config.props https://github.com/NuGet/NuGet.Client/blob/dev/build/config.props#L15-L18. Whereas for [optprof pipeline](https://devdiv.visualstudio.com/DevDiv/_build?definitionId=16906&_a=summary) the build number is automatically generated by Azure DevOps.

---

> Whereas for [optprof pipeline](https://devdiv.visualstudio.com/DevDiv/_build?definitionId=16906&_a=summary) the build number is automatically generated by Azure DevOps.

Build number is actually set by a task in the OptProf pipeline. https://github.com/NuGet/NuGet.Client/blob/dev/eng/pipelines/optprof.yml#L73-L78. I added `Build.BuildNumber` value to the build number just to be able to run OptProf pipeline for the same official build. It was not the case earlier because `Upload VSTS drop task` failed with `Drop Already Exists` exception in my testing.

> However, I'm not sure what other consequences there are. 
I tried changing the prefix to `OptProf` but the [`Upload VSTS drop` task failed](https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=6908721&view=logs&j=87ee994e-8d25-5c3a-5836-c28a1c54ff14&t=12c880cb-a8ee-5c18-26cb-685efba1b642)  with `##[error]Cannot find path 'C:\Users\CLOUDT~1\AppData\Local\Temp\OptProf\DevDiv\NuGet-NuGet.Client-Trusted\dev-kartheekp-ms-differentvsbootstrappers-useOptProfForBuildNumber\dev-kartheekp-ms-differentvsbootstrappers-useOptProfForBuildNumber_7de18ecea5607b9d34cef4d4dd4f7074cbc8d1ca' because it does not exist.` error

I think at this point it is safe to merge this PR because build number of OptProf pipeline will not clash with Official build.

---

> I tried changing the prefix to OptProf but the Upload VSTS drop task failed with

If you read the log, it says that the drop finalization was complete, and further up there's a link to "view the drop from a browser", where we can see all the files are there.

I'm guessing that the error happened when the task is cleaning up after itself, and you'll notice that the path it listed exceeds Windows' MAX_PATH, due to the extremely long branch name and build number used. My bet is that the machine either doesn't have long path enabled, or the process that the task is running doesn't have the right win32 manifest. Both are necessary for long paths to work, it's not enough to only enable the registry key ([for example](https://github.com/NuGet/NuGet.Client/pull/4830)).

> This build has Build.BuildNumber appended to build name

sorry to be pedantic, just trying to make sure you understand how the build works in case it helps reduce confusion in the future. There is no variable "build name". The script further up overwrites BuildNumber, at which point the previous BuildNumber is no longer available. Your change was to append the auto-generated BuildNumber in the new value for BuildNumber.

I find it really unfortunate that the MicroBuildBuildVSBootstrapper task doesn't appear to have any parameters to change the URL pattern it uses. This pipeline is a bit "special", since this build itself is not the important thing, the "component under test" is. As a consequence it would be ideal for us to keep the BuildNumber "clean" for when viewing the list of builds the pipeline has run, but we need to use a different, unique number just for the drop URL.